### PR TITLE
restore workflow crud api

### PR DESCRIPTION
**Pull request by danielnorberg**
* This reverts PR #255 (3aac3da) and reintroduces workflow CRUD REST API endpoints.

* The Styx workflow storage is now the source of truth.

* Integrations with CD platform and ingestion from external truth sources can be performed directly using the workflow CRUD REST API endpoints.

* Unavailability of CD platforms and external truth sources does not affect the workflows that are stored by Styx itself.

### CODE REVIEW

File: doc/api.apib [hunk 1]
```diff
@@ -102,6 +102,75 @@ Resources related to the Workflows registered in Styx.
               }
         }
 
+### Delete Workflow [DELETE]
+
+Returns 200 OK if the Workflow was found an deleted, 404 Not Found otherwise.
```
Comment (honnix) [inline doc/api.apib]: There is now `204`.

Comment (danielnorberg) [reply doc/api.apib]: good catch!

**Comment by bergman** [comment]
Looks good, but what about validation of input? I expected some `dockerImageValidator` calls in the POST too, not only the PATCH.

**Comment by danielnorberg** [comment]
@bergman good catch!

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java [hunk 2]
```diff
@@ -62,10 +67,14 @@
   public SchedulerResource(
       StateManager stateManager,
       TriggerListener triggerListener,
+      Consumer<Workflow> workflowChangeListener,
+      Consumer<Workflow> workflowRemoveListener,
       Storage storage,
       Time time) {
     this.stateManager = Objects.requireNonNull(stateManager);
     this.triggerListener = Objects.requireNonNull(triggerListener);
+    this.workflowChangeListener = workflowChangeListener;
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: missing null check.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 3]
```diff
@@ -86,11 +94,61 @@ public WorkflowResource(Storage storage, DockerImageValidator dockerImageValidat
...
+      return CompletableFuture.completedFuture(
+          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Missing payload.")));
+    }
+    final WorkflowConfiguration workflowConfig;
+    try {
+      workflowConfig = OBJECT_MAPPER
+          .readValue(payload.get().toByteArray(), WorkflowConfiguration.class);
+    } catch (IOException e) {
+      return CompletableFuture.completedFuture(
+          Response.forStatus(Status.BAD_REQUEST.withReasonPhrase("Invalid payload.")));
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: We can improve the error message, I think. Instead of `Invalid payload` we can be more specific, like `Invalid workflow configuration` for example.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java [hunk 4]
```diff
@@ -86,11 +94,61 @@ public WorkflowResource(Storage storage, DockerImageValidator dockerImageValidat
...
+    }
+
+    final Workflow workflow = Workflow.create(componentId, workflowConfig);
+
+    final ByteString requestPayload;
+    try {
+      requestPayload = serialize(workflow);
+    } catch (JsonProcessingException e) {
+      return CompletableFuture.completedFuture(Response.forStatus(
+          Status.INTERNAL_SERVER_ERROR.withReasonPhrase("Failed to serialize proxy payload.")));
```
Comment (honnix) [inline styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Let's get away from the `proxy` since it is completely internal to how we implement it.

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java [hunk 5]
```diff
@@ -82,11 +91,42 @@ public SchedulerResource(
...
+    }
+    if (!workflowOpt.isPresent()) {
+      return Response.forStatus(Status.NOT_FOUND.withReasonPhrase("Workflow does not exist"));
+    }
+    workflowRemoveListener.accept(workflowOpt.get());
+    return Response.forStatus(Status.NO_CONTENT);
+  }
+
+  private Response<Workflow> createOrUpdateWorkflow(Workflow workflow) {
+    workflowChangeListener.accept(workflow);
```
Comment (honnix) [inline styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: It's better to have a big `TODO` here: validate the workflow and throw something more reasonable than 500 if `accept` failed with `RuntimeException`. This can be address in a different PR to keep this one clean.

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 3 above)
Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: Might be completely unparsable payload though... 🤔

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 4 above)
Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: lol how did this go through the first PR

Comment (danielnorberg) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: also why is it re-serializing instead of just passing on inputs...

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 3 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: We need a flaky network for this to happen...

File: styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java (hunk 4 above)
Comment (honnix) [reply styx-api-service/src/main/java/com/spotify/styx/api/WorkflowResource.java]: not sure. I thought we fixed this.

**Comment by danielnorberg** [comment]
Btw, the use of `POST` for upsert is a bit unusual. `PUT` is usually used for that. Do we care?

File: styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java (hunk 5 above)
Comment (danielnorberg) [reply styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: We should validate the workflow as much as possible, and return 4XX for bad requests, but if something fails after validation I think that's an internal error that should be surfaced as a 500.

Comment (honnix) [reply styx-scheduler-service/src/main/java/com/spotify/styx/api/SchedulerResource.java]: There was a corner case that invalid workflows are surfaced later when parsing cron syntax to get next natural trigger. We can address that in a separated PR.

**Comment by honnix** [comment]
If using PUT I think we will need to include workflow Id in the url.

**Comment by honnix** [comment]
Maybe not worth it.

**Comment by danielnorberg** [comment]
Yeah, PUT would entail having workflow id in the url. I'm fine with POST for now though. If we want we can change it later.

**Comment by danielnorberg** [comment]
Good to merge?

**Comment by honnix** [comment]
Sorry for being late. Sure. 👍
